File: poly/update/update.fs

#light

open System
open System.IO
open System.Diagnostics


let ops = [
    (".bzr", ["bzr pull"; "bzr update"])
    (".hg",  ["hg pull"; "hg update"])
    (".svn", ["svn update"])
    (".git", ["git pull"])] |> Map.ofList

let srcdir = "/home/sa/src"

let run (cmd : string) =
    let args = cmd.Split()
    let x, y = args.[0], args.[1]
    Process.Start(x, y).WaitForExit()


let update path = 
    let get_dirs path = Array.toList(Directory.GetDirectories(path))
    let projects = get_dirs path
    for proj in projects do
        let dirs = get_dirs proj
        for dir in dirs do
            let dirname = Path.GetFileName(dir)
            if ops.ContainsKey(dirname) then
                printfn "\nproj: %A" proj
                Directory.SetCurrentDirectory(proj)
                for cmd in ops.Item(dirname) do
                    run cmd |> ignore

[<EntryPoint>]
let main args = 
    //printfn "args: %A" args
    if args.Length > 0 then Array.iter update args
    else update srcdir
    0

File: poly/update/update.cs
using System;
using System.IO;
using System.Diagnostics;

class UpdateManager {
	public static void runProcess(String name, String args) {
		Process proc = new Process();
		 proc.EnableRaisingEvents = false;
		 proc.StartInfo.FileName = name;
		 proc.StartInfo.Arguments = args;
		 proc.Start();
		 proc.WaitForExit();
	} 
    
    public static void scanDirectory(string path) {
		Console.WriteLine("scanning: {0}\n", path);

		foreach(String project in Directory.GetDirectories(path)) {
			Console.WriteLine("\nupdating: {0}", project);

			foreach(String dir in Directory.GetDirectories(project)) {
				String directory = Path.GetFileName(dir);

				Directory.SetCurrentDirectory(project);

				switch (directory) {
				case ".bzr":
					runProcess("bzr", "pull");
					runProcess("bzr", "update");
					break;

				case ".svn":
					runProcess("svn", "update");
					break;

				case ".git":
					runProcess("git", "pull");
					break;

				case ".hg":
					runProcess("hg", "pull");
					runProcess("hg", "update");
					break;

				default:
					break;
				}

			}

		}

	}

	public static void Main(string[]args) {
		String PATH = "/home/sa/src";

		if (args.Length >= 1) {
			foreach(string arg in args) {
				scanDirectory(arg);
			}
		} else {
			scanDirectory(PATH);
		}
	}
}

